### ISSUE TITLE
delay before take_off

### ISSUE THREAD

**Issue by MaximVorotnikov** [mavlink/MAVSDK#2714]
I'm working with the ArduPilot firmware. I want to make a delay after arm so that there is a delay between arm and takeoff. The problem is that I can do it every other time. From time to time, the takeoff function returns Failed to me, while the motors start to spin harder and the drone either takes off a few centimeters or even tries to turn over. The question is why if you send arm and take_off without delay, then everything works fine, but if you add a delay, then problems arise. Here is a sample code where I add a delay.

class DroneController : public rclcpp::Node {
public:
    DroneController() : Node("drone_controller"), 
                       mavsdk{mavsdk::Mavsdk::Configuration{mavsdk::Mavsdk::ComponentType::GroundStation}} {
        
        std::string connection_url = "serial:///dev/ttyTHS1:921600";
        mavsdk::ConnectionResult connection_result = mavsdk.add_any_connection(connection_url);
        
        if (connection_result != mavsdk::ConnectionResult::Success) {
            RCLCPP_ERROR(this->get_logger(), "Error connection");
            return;
        }

        RCLCPP_INFO(this->get_logger(), "waiting connection of drone...");
        

        auto system_result = mavsdk.first_autopilot(10.0);
        if (!system_result) {
            RCLCPP_ERROR(this->get_logger(), "timeout wait autopilot");
            return;
        }
        system = system_result;
        

        action = std::make_shared<mavsdk::Action>(system.value());
        telemetry = std::make_shared<mavsdk::Telemetry>(system.value());

        action_server_ = rclcpp_action::create_server<Commands>(
            this,
            "drone_commands",
            std::bind(&DroneController::handle_goal, this, std::placeholders::_1, std::placeholders::_2),
            std::bind(&DroneController::handle_cancel, this, std::placeholders::_1),
            std::bind(&DroneController::handle_accepted, this, std::placeholders::_1)
        );

        RCLCPP_INFO(this->get_logger(), "drone ready");
    }

private:
    mavsdk::Mavsdk mavsdk;
    std::optional<std::shared_ptr<mavsdk::System>> system;
    std::shared_ptr<mavsdk::Action> action;
    std::shared_ptr<mavsdk::Telemetry> telemetry;
    rclcpp_action::Server<Commands>::SharedPtr action_server_;

    rclcpp_action::GoalResponse handle_goal(
        const rclcpp_action::GoalUUID& uuid,
        std::shared_ptr<const Commands::Goal> goal) {
        RCLCPP_INFO(this->get_logger(), "A goal request was received from action_type: %s", goal->action_type.c_str());
        (void)uuid;
        return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
    }

    rclcpp_action::CancelResponse handle_cancel(
        const std::shared_ptr<GoalHandleCommands> goal_handle) {
        RCLCPP_INFO(this->get_logger(), "Received a request to cancel a goal");
        (void)goal_handle;
        return rclcpp_action::CancelResponse::ACCEPT;
    }

    void handle_accepted(const std::shared_ptr<GoalHandleCommands> goal_handle) {
        std::thread{std::bind(&DroneController::execute, this, std::placeholders::_1), goal_handle}.detach();
    }

    void execute(const std::shared_ptr<GoalHandleCommands> goal_handle) {
        const auto goal = goal_handle->get_goal();
        auto result = std::make_shared<Commands::Result>();

        RCLCPP_INFO(this->get_logger(), "Completing a goal with action_type: %s", goal->action_type.c_str());

        if (!system) {
            RCLCPP_ERROR(this->get_logger(), "The system is not initialized");
            result->success = false;
            goal_handle->abort(result);
            return;
        }

        if (goal->action_type == "arm") 
        {
            const mavsdk::Action::Result arm_result = action->arm();

            if (arm_result != mavsdk::Action::Result::Success) 
            {
                std::cerr << "Arming failed: " << arm_result << '\n';
                result->success = false;
                goal_handle->abort(result);

            }

        } 
        else if (goal->action_type == "take_off_after_arm") 
        {
            action->set_takeoff_altitude(goal->alt_takeoff);

            std::cout << "Taking off...\n";
            mavsdk::Action::Result takeoff_result = action->takeoff();
            if (takeoff_result != mavsdk::Action::Result::Success) 
            {
                std::cerr << "Takeoff failed: " << takeoff_result << '\n';
                result->success = false;
                goal_handle->abort(result);
            }

        } 
        else if (goal->action_type == "take_off_full_cycle") 
        {
            const mavsdk::Action::Result arm_result = action->arm();

            if (arm_result != mavsdk::Action::Result::Success) 
            {
                std::cerr << "Arming failed: " << arm_result << '\n';
                result->success = false;
                goal_handle->abort(result);

            }
            int time = goal->loiter_time_after_arm * 1000;
            std::this_thread::sleep_for(std::chrono::milliseconds(time));

            action->set_takeoff_altitude(goal->alt_takeoff);

            std::cout << "Taking off...\n";
            mavsdk::Action::Result takeoff_result = action->takeoff();
            if (takeoff_result != mavsdk::Action::Result::Success) 
            {
                std::cerr << "Takeoff failed: " << takeoff_result << '\n';
                result->success = false;
                goal_handle->abort(result);
            }

        } 
        else if (goal->action_type == "land") 
        {
            std::cout << "Landing...\n";
            const mavsdk::Action::Result land_result = action->land();
            if (land_result != mavsdk::Action::Result::Success) {
                std::cerr << "Land failed: " << land_result << '\n';
                result->success = false;
                goal_handle->abort(result);
            }

            // Check if vehicle is still in air
            while (telemetry->in_air()) {
                std::cout << "Vehicle is landing...\n";
                std::this_thread::sleep_for(std::chrono::seconds(1));
            }
            std::cout << "Landed!\n";
        } 

        else 
        {
            RCLCPP_ERROR(this->get_logger(), "unknow type action: %s", goal->action_type.c_str());
            result->success = false;
            goal_handle->abort(result);
        }
    }
};

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<DroneController>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

that is, my server accepts "arm" and then "take_off_after_arm" or immediately accepts "take_off_full_cycle". Both the first and second scenarios work every other time. Why is this happening? Perhaps this delay before takeoff should be set by some parameter, but I did not find it in the ArduPilot firmware settings. Please help me.

**Comment by julianoes** [comment]
What mode are you in when you do this? And have you asked in the ArduPilot forum or Discord? I'd expect people to be able to help you there. And it doesn't sound like a mavsdk problem as such, so I'm not sure what to say.

**Comment by MaximVorotnikov** [comment]
I installed MAVSDK from the v3 branch and everything works successfully, before that, another developer installed MAVSDK, switching for some reason to the v2.12.2 tag, which caused this problem. But although it is strange for me personally, before that the drone could take off in any mode, now it is on this version, if it is in Guided mode, then it only makes an arm and does not take off, and if you change the mode to Stabilized before that, then the takeoff is successful.

**Comment by julianoes** [comment]
Ok thanks for the feedback. I'll close this issue then.
